server/src/main/java/org/elasticsearch/index/engine/TranslogDirectoryReader.java

```diff
@@ -76,21 +80,48 @@
  * into an in-memory Lucene segment that is created on-demand.
  */
 final class TranslogDirectoryReader extends DirectoryReader {
-    private final TranslogLeafReader leafReader;
+    private final LeafReader leafReader;
 
-    TranslogDirectoryReader(
+    static DirectoryReader create(
         ShardId shardId,
         Translog.Index operation,
         MappingLookup mappingLookup,
         DocumentParser documentParser,
         EngineConfig engineConfig,
         Runnable onSegmentCreated
     ) throws IOException {
-        this(new TranslogLeafReader(shardId, operation, mappingLookup, documentParser, engineConfig, onSegmentCreated));
+        final Directory directory = new ByteBuffersDirectory();
+        boolean success = false;
+        try {
+            final LeafReader leafReader;
+            // When using synthetic source, the translog operation must always be reindexed into an in-memory Lucene to ensure consistent
+            // output for realtime-get operations. However, this can degrade the performance of realtime-get and update operations.
```

Comment:
The inconsistencies would those purely be white-space and field order or also actual content like order in arrays etc?

I *think* this is ok, synthetic source and heavy real-time GET / update use cases are probably rare, but I'd like to understand the tradeoff more deeply.

If it is only white-space and field order, could we sort that instead and how difficult would that be? Not asking to implement that now, more about knowing our backdoor in case we need this.

Reply:
Yes, the inconsistencies also include removing duplicate values in an array and returning a single value as an object instead of an array. For example:

```
{ 
	"f": ["v"]
}
```

becomes

```
{
	"f": "v"
}
```

Or:

```
{
	"f": ["v1", "v1", "v2"]
}
```

becomes

```
{
	"f": ["v1", "v2"]
}
```